modules/core/src/main/java/org/apache/ignite/hlc/HybridTimestamp.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.hlc;
+
+import org.apache.ignite.internal.tostring.S;
+
+/**
+ * A hybrid timestamp that combines physical clock and logical clock.
+ */
+public class HybridTimestamp implements Comparable<HybridTimestamp> {
```

Comment:
The implementation seems correct, but not optimal. It produces new objects and does unnecessary operations. I suggest to look at https://github.com/eventreducer/hlc/blob/master/src/main/java/org/eventreducer/hlc/HybridTimestamp.java as a good reference. It also has a nice abstaction of physical time provider.